kyo-combinators/shared/src/main/scala/kyo/KYO.scala

```diff
@@ -0,0 +1,209 @@
+
+    def fromAutoCloseable[A <: AutoCloseable, S](closeable: => A < S): A < (S & Resources) =
+        acquireRelease(closeable)(c => IOs(c.close()))
+
+    def fromEither[E, A](either: => Either[E, A]): A < Aborts[E] =
+        Aborts.get(either)
+
+    def fromFuture[A: Flat, S](future: => Future[A] < S): A < (S & Fibers) =
+        future.map(f => Fibers.fromFuture(f))
+
+    def fromPromiseScala[A: Flat, S](promise: => scala.concurrent.Promise[A] < S)
+        : A < (S & Fibers) =
+        promise.map(p => fromFuture(p.future))
+
+    def fromOption[A, S](option: => Option[A] < S): A < (S & Options) =
+        Options.get(option)
+
+    def fromSeq[A, S](sequence: => Seq[A] < S): A < (S & Choices) =
+        sequence.flatMap(seq => Choices.get(seq))
+
+    def fromTry[A, S](_try: => scala.util.Try[A] < S): A < (S & Aborts[Throwable]) =
+        _try.map(t => Aborts.get(t.toEither))
+
+    inline def logInfo[S](message: => String < S): Unit < (S & IOs) =
+        message.map(m => Logs.info(m))
+
+    inline def logInfo[S, S1](
+        message: => String < S,
+        err: => Throwable < S1
+    ): Unit < (S & S1 & IOs) =
+        message.map(m => err.map(e => Logs.info(m, e)))
+
+    inline def logWarn[S](message: => String < S): Unit < (S & IOs) =
+        message.map(m => Logs.warn(m))
+
+    inline def logWarn[S, S1](
+        message: => String < S,
+        err: => Throwable < S1
+    ): Unit < (S & S1 & IOs) =
+        message.map(m => err.map(e => Logs.warn(m, e)))
+
+    inline def logDebug[S](message: => String < S): Unit < (S & IOs) =
+        message.map(m => Logs.debug(m))
+
+    inline def logDebug[S, S1](
+        message: => String < S,
+        err: => Throwable < S1
+    ): Unit < (S & S1 & IOs) =
+        message.map(m => err.map(e => Logs.debug(m, e)))
+
+    inline def logError[S](message: => String < S): Unit < (S & IOs) =
+        message.map(m => Logs.error(m))
+
+    inline def logError[S, S1](
+        message: => String < S,
+        err: => Throwable < S1
+    ): Unit < (S & S1 & IOs) =
+        message.map(m => err.map(e => Logs.error(m, e)))
+
+    inline def logTrace[S](message: => String < S): Unit < (S & IOs) =
+        message.map(m => Logs.trace(m))
+
+    inline def logTrace[S, S1](
+        message: => String < S,
+        err: => Throwable < S1
+    ): Unit < (S & S1 & IOs) =
+        message.map(m => err.map(e => Logs.trace(m, e)))
+
+    def never: Nothing < Fibers =
+        Fibers.never.join
+            *> IOs(throw new IllegalStateException("Fibers.never completed"))
+
+    val none: Nothing < Options =
+        Options.empty
+
+    def provide[D, SD, A, SA, E, SR](
+        dependency: => D < SD
+    )(
+        effect: A < (SA & Envs[E])
+    )(
+        using
+        he: HasEnvs[D, E] { type Remainder = SR },
+        t: Tag[D],
+        fl: Flat[A]
+    ): A < (SA & SD & SR) =
+        dependency.map(d => Envs.run[D, A, SA, E, SR](d)(effect))
+        // V >: Nothing: Tag, T: Flat, S, VS, VR
```

Comment:
?

Reply:
That was just to help me work out what type parameters to use. Didn't mean to leave it there.